// gulp-jasmine-browser/index.js
'use strict';

exports.__esModule = true;
exports.specRunner = exports.chrome = exports.phantomjs = exports.slimerjs = exports.server = exports.headless = undefined;

var _through = require('through2');

var _headless = require('./lib/headless');

var _spec_runner = require('./lib/spec_runner');

var _spec_runner2 = _interopRequireDefault(_spec_runner);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function specRunner() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      profile = _ref.profile,
      sourcemappedStacktrace = _ref.sourcemappedStacktrace;

  var specRunner = new _spec_runner2.default({ profile: profile, sourcemappedStacktrace: sourcemappedStacktrace, path: '/specRunner.html' });
  var consoleRunner = new _spec_runner2.default({ console: true, path: '/consoleRunner.html' });
  return (0, _through.obj)(function (file, encoding, next) {
    this.push(file);
    this.push(specRunner.addFile(file.relative));
    this.push(consoleRunner.addFile(file.relative));
    next();
  });
}

exports.headless = _headless.headless;
exports.server = _headless.server;
exports.slimerjs = _headless.slimerjs;
exports.phantomjs = _headless.phantomjs;
exports.chrome = _headless.chrome;
exports.specRunner = specRunner;
